### PR TITLE
Reflective setAccessible(true) will produce scary warnings on the con…

### PR DESCRIPTION
**Pull request by normanmaurer**
…sole when using java9+, dont do it

Motivation:

Reflective setAccessible(true) will produce scary warnings on the console when using java9+, while netty still works. That said users may feel uncomfortable with these warnings, we should not try to do it by default when using java9+.

Modifications:

Add io.netty.noReflectiveAccessible system property which controls if setAccessible(...) will be used. By default it will bet set to true when using java9+.

Result:

Fixes [#7254].

### CODE REVIEW

**Comment by normanmaurer** [comment]
@doom369 FYI...

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java [hunk 1]
```diff
@@ -84,7 +86,9 @@
                 public Object run() {
                     try {
                         final Field unsafeField = Unsafe.class.getDeclaredField("theUnsafe");
-                        Throwable cause = ReflectionUtil.trySetAccessible(unsafeField);
+                        // We always want to try using Unsafe as the access still works on java9 as well and
+                        // we need it for out native-transports and many optimizations.
```
Comment (johnou) [inline common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: @normanmaurer have you looked into MethodHandles.byteBufferViewVarHandle [1] that creates a VarHandle that allows to do aligned/unaligned access on a ByteBuffer?

[1] http://download.java.net/java/jdk9/docs/api/java/lang/invoke/MethodHandles.html#byteBufferViewVarHandle-java.lang.Class-java.nio.ByteOrder-

http://mail.openjdk.java.net/pipermail/jigsaw-dev/2017-February/011477.html

**Comment by doom369** [comment]
@normanmaurer 

`java.nio.DirectByteBuffer(long,int)` and `java.nio.Bits.unaligned()` both are not available under Java 9. So maybe it would be better just to skip reflection access try for those particular methods when running java9 instead of adding the additional property?

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java (hunk 1 above)
Comment (normanmaurer) [reply common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: @johnou we are compiling against java8 so it will not help at all (and most people not use java9 yet).

**Comment by normanmaurer** [comment]
@doom369 the property will be "true" by default when using java9 and so we will skip it, or I am missing anything? The thing is you can open it up if you like to and so get the extra perf / features. Thats why I would love to at least still allow this.

**Comment by doom369** [comment]
@normanmaurer I see... In case someone manually (via jvm option) will open access to jdk internal modules this will work. Doesn't seem like real case to me, but yeah. Nice to have.

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java (hunk 1 above)
Comment (johnou) [reply common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: @normanmaurer perhaps for Netty 5 then given that Java 8 and 9 will be EOL this year unless you fork out cash http://www.oracle.com/technetwork/java/eol-135779.html

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java [hunk 2]
```diff
@@ -797,6 +801,15 @@ private static boolean isAndroid0() {
         return android;
     }
 
+    private static boolean explicitNoReflectiveAccessible0() {
+        // we disable reflective access
+        return SystemPropertyUtil.getBoolean("io.netty.noReflectiveAccessible", javaVersion() >= 9);
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: can we avoid the negative "no" in the name? IMHO it makes it more difficult to have to negate the logic to understand what is going on. What about `io.netty.tryReflectionSetAccessible`?

Comment (normanmaurer) [reply common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: @Scottmitch sure... I just used it because we use it for unsafe as well

**Comment by normanmaurer** [comment]
@Scottmitch PTAL again

**Comment by normanmaurer** [comment]
@doom369 I think it is definitely a real-case if you want max perf.

**Comment by ejona86** [comment]
CC @zpencer

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java [hunk 3]
```diff
@@ -46,6 +46,8 @@
 
     private static final Throwable UNSAFE_UNAVAILABILITY_CAUSE;
     private static final Object INTERNAL_UNSAFE;
+    private static final boolean IS_EXPLICIT_TRY_REFLECTIVION_SET_ACCESSIBLE = explicitTryReflectionSetAccessible0();
```
Comment (zpencer) [inline common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: nit: typo in REFLECTION/REFLECTIVE

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java [hunk 4]
```diff
@@ -84,7 +86,9 @@
                 public Object run() {
                     try {
                         final Field unsafeField = Unsafe.class.getDeclaredField("theUnsafe");
-                        Throwable cause = ReflectionUtil.trySetAccessible(unsafeField);
+                        // We always want to try using Unsafe as the access still works on java9 as well and
+                        // we need it for out native-transports and many optimizations.
+                        Throwable cause = ReflectionUtil.trySetAccessible(unsafeField, false);
```
Comment (zpencer) [inline common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: Would the scary sounding message still be logged here for jdk9? Perhaps we can log a message for the user in addition to the code comments here.

Comment (normanmaurer) [reply common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: Nope... access to unsafe is handled special and so nothing is logged

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (e72c197aa3a88e97c10793ee7870775ab8807f92) and 4.0 (c62345f3a9)

**Comment by normanmaurer** [comment]
Yep with JNI you can do whatever you want 👌😉

**Comment by eolivelli** [comment]
@normanmaurer 
just to double check,
after upgrading Netty to this version (4.1.21Final) is it possible to have worse performance on Java 9 without setting the system property an re-allow reflective access ?

**Comment by normanmaurer** [comment]
@eolivelli yes... that said there is not much we can do about it :(

**Comment by eolivelli** [comment]
@normanmaurer clear, thank you

**Comment by eolivelli** [comment]
@normanmaurer  last thought: can you consider adding a note here https://netty.io/news/2018/02/05/4-0-56-Final-4-1-21-Final.html I think this is an important flag to set and people may see degradation in performances.

**Comment by slandelle** [comment]
@normanmaurer Wouldn't it make sense to ask the JCP for a public API for what you're doing?

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java (hunk 1 above)
Comment (johnou) [reply common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: @slandelle like this suggestion or is the public API not rich enough?

**Comment by artash** [comment]
Just a side note here, I think this change _may_ affect the RSS of your java process. Your direct buffers will now be allocated with `ByteBuffer.allocateDirect()` and this will not only allocate the memory but also set it to zero, i.e. it'll call:
```
            base = UNSAFE.allocateMemory(size);
``` 
followed by 
```
        UNSAFE.setMemory(base, size, (byte) 0);
```
 whereas in the past netty would only call `Unsafe.allocateMemory()` and not set the memory.  This means the memory would be committed but won't be resident (in physical RAM) until it's written to. 

In other words the new version eagerly uses all the memory allocated and this may definitely have an impact in your free memory graphs.

**Comment by itsTyrion** [comment]
> @eolivelli yes... that said there is not much we can do about it :(

Any, even vague, numbers? 1%, 5%, 10%, 20%?

**Comment by itsTyrion** [comment]
@normanmaurer
